numpy/core/src/multiarray/item_selection.c

```diff
@@ -1749,6 +1750,263 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
+            PyDataMem_FREE(indbuffer);
+            PyErr_NoMemory();
+            goto fail;
+        }
+        while (size--) {
+            for (i = 0; i < N; i++) {
+                indbuffer[i] = i;
+            }
+            for (j = 0; j < n; j++) {
+                int rcode;
+                elsize = PyArray_DESCR(mps[j])->elsize;
+                astride = PyArray_STRIDES(mps[j])[axis];
+                argsort = PyArray_DESCR(mps[j])->f->argsort[NPY_STABLESORT];
+                if(argsort == NULL) {
+                    argsort = npy_atimsort;
+                }
+                _unaligned_strided_byte_copy(valbuffer, (npy_intp) elsize,
+                                             its[j]->dataptr, astride, N, elsize);
+                if (PyArray_ISBYTESWAPPED(mps[j])) {
+                    _strided_byte_swap(valbuffer, (npy_intp) elsize, N, elsize);
+                }
+                rcode = argsort(valbuffer, indbuffer, N, mps[j]);
+                if (rcode < 0 || (PyDataType_REFCHK(PyArray_DESCR(mps[j]))
+                            && PyErr_Occurred())) {
+                    PyDataMem_FREE(valbuffer);
+                    PyDataMem_FREE(indbuffer);
+                    goto fail;
+                }
+                PyArray_ITER_NEXT(its[j]);
+            }
+            char *slicebuffer = PyDataMem_NEW(N*self_elsize);
+            _unaligned_strided_byte_copy(slicebuffer, self_elsize, sit->dataptr,
+                                         self_stride, N, self_elsize);
+
+            for(i = 0; i < N; i++){
+                int index = indbuffer[i];
+                char *startOld = slicebuffer + (index * self_elsize);
+                memmove(sit->dataptr + (i*self_stride),
+                            startOld, self_elsize);
+            }
+            PyArray_ITER_NEXT(sit);
+            PyDataMem_FREE(slicebuffer);
+        }
+        PyDataMem_FREE(valbuffer);
+        PyDataMem_FREE(indbuffer);
+    }
+    else {
+        while (size--) {
+            for (i = 0; i < N; i++) {
+                indbuffer[i] = i;
+            }
+            for (j = 0; j < n; j++) {
+                int rcode;
+                argsort = PyArray_DESCR(mps[j])->f->argsort[NPY_STABLESORT];
+                if(argsort == NULL) {
+                    argsort = npy_atimsort;
+                }
+                rcode = argsort(its[j]->dataptr,
+                        indbuffer, N, mps[j]);
+                if (rcode < 0 || (PyDataType_REFCHK(PyArray_DESCR(mps[j]))
+                            && PyErr_Occurred())) {
+                    goto fail;
+                }
+
+                PyArray_ITER_NEXT(its[j]);
+            }
+            char *sortedbuffer = PyDataMem_NEW(N*self_elsize);
+            if (sortedbuffer == NULL) {
+                PyDataMem_FREE(indbuffer);
+                PyErr_NoMemory();
+                goto fail;
+            }
+            for(i = 0; i < N; i++){
+                int index = indbuffer[i];
```

Comment:
This must be `npy_intp`, compiler shoudl warn about it though?

Reply:
Same in the buffered version above.

Reply:
Ah i agree. will fix it